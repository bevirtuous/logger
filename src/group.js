import { logger } from './logger';

const KEY_COLOR = 'gray';
const FONT_WEIGHT_NORMAL = 'lighter';
const FONT_WEIGHT_BOLD = 'bold';

const defaultStyles = {
  color: KEY_COLOR,
  weight: FONT_WEIGHT_NORMAL,
};
const defaultDate = new Date();

/**
 * Repeates a string by a defined amount of times.
 * @param {string} str The string to repeat.
 * @param {number} times How many times the string should be repeated.
 * @return {string}
 */
export function repeat(str, times) {
  if (!times) {
    return str;
  }

  return (new Array(times + 1)).join(str);
}

/**
 * Creates a string from a number and fills up the gap with 0 to reach a certain length.
 * @param {number} num The number
 * @param {number} maxLength The length of the resulting string.
 * @return {string}
 */
export function pad(num, maxLength) {
  if (!maxLength || num.toString().length === maxLength) {
    return `${num}`;
  }

  return repeat('0', maxLength - num.toString().length) + num;
}

/**
 * @param {Object} [time] The base time.
 * @return {string}
 */
export function getFormattedTime(time = defaultDate) {
  const hours = pad(time.getHours(), 2);
  const minutes = pad(time.getMinutes(), 2);
  const seconds = pad(time.getSeconds(), 2);
  const milliseconds = pad(time.getMilliseconds(), 3);

  return `${hours}:${minutes}:${seconds}.${milliseconds}`;
}

/**
 * Returns the biggest length value of an object's keys.
 * @param {Object} prop The object to walk through.
 * @return {number}
 */
export function maxKeysLength(prop) {
  let maxLength = 0;

  Object.keys(prop).forEach((key) => {
    if (key.length > maxLength) {
      maxLength = key.length;
    }
  });

  return maxLength;
}

/**
 * @param {Object} props The styling props.
 * @param {string} [props.color='gray'] The text color.
 * @param {string} [props.weight='lighter'] The font weight.
 * @return {string}
 */
export function style({ color, weight } = defaultStyles) {
  return `color: ${color}; font-weight: ${weight};`;
}

/**
 *
 * @param {string} title The title of the log group.
 * @param {Object} [content={}] Some content to show in the group.
 * @param {string} [color='inherit'] The color for the title.
 */
export function group(title, content = {}, color = 'inherit') {
  const time = getFormattedTime();

  if (!Object.keys(content).length) {
    logger.log(` ${title} @ ${time}`);
    return;
  }

  logger.groupCollapsed(
    ` %c${title} %c@ ${time}`,
    style({
      color,
      weight: FONT_WEIGHT_NORMAL,
    }),
    style({
      color: 'inherit',
      weight: FONT_WEIGHT_BOLD,
    }),
    style()
  );

  const maxLength = maxKeysLength(content) + 2;

  Object.keys(content).forEach((key) => {
    const value = content[key];
    const action = `${key}:`.padEnd(maxLength);
    const styles = {
      color: KEY_COLOR,
      weight: FONT_WEIGHT_BOLD,
    };

    // If the content is an object, log all keys individually.
    if (typeof value === 'object' && value !== null && value.constructor === Object) {
      if (!Object.keys(value).length) {
        logger.log(` %c ${action}`, style(styles), undefined);
      } else {
        logger.log(` %c ${action}`, style(styles), value);
      }
    } else {
      logger.log(` %c ${action}`, style(styles), value);
    }
  });

  logger.groupEnd();
}
