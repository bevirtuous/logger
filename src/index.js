export * from './logger';
export * from './group';
