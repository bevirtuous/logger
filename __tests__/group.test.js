import * as group from '../src/group';
import { logger } from '../src/logger';

describe('group', () => {
  describe('repeat', () => {
    it('returns the original string if no repeantance defined', () => {
      const result = group.repeat('foo');
      expect(result).toBe('foo');
      expect(result).toMatchSnapshot();
    });

    it('returns the string 5 times', () => {
      const result = group.repeat('foo', 5);
      expect(result).toBe('foofoofoofoofoo');
      expect(result).toMatchSnapshot();
    });
  });

  describe('pad', () => {
    it('returns a string of a number with one leading 0', () => {
      const result = group.pad(5);
      expect(result).toBe('5');
      expect(result).toMatchSnapshot();
    });

    it('returns a string of a number wuth three leading 0', () => {
      const result = group.pad(5, 4);
      expect(result).toBe('0005');
      expect(result).toMatchSnapshot();
    });
  });

  describe('getFormattedTime', () => {
    const time = new Date(Date.UTC(2018, 7, 30, 22, 38, 38, 396));
    it('returns a formatted time string', () => {
      const result = group.getFormattedTime(time);
      expect(result).toBe('00:38:38.396');
      expect(result).toMatchSnapshot();
    });
  });

  describe('maxKeysLength', () => {
    it('should return the highest length of a key in an object', () => {
      const result = group.maxKeysLength({
        foo: 1,
        foobar: 1,
        bar: 1,
      });
      expect(result).toBe(6);
      expect(result).toMatchSnapshot();
    });
  });

  describe('style', () => {
    it('should return a valid styling with custom values.', () => {
      const result = group.style({
        color: '#fff',
        weight: 'bold',
      });
      expect(result).toBe('color: #fff; font-weight: bold;');
      expect(result).toMatchSnapshot();
    });

    it('should return a valid styling with default values.', () => {
      const result = group.style();
      expect(result).toBe('color: gray; font-weight: lighter;');
      expect(result).toMatchSnapshot();
    });
  });

  describe('group', () => {
    it('logs a simple log if not options are set', () => {
      logger.log = jest.fn();
      group.group('some title');
      expect(logger.log).toHaveBeenCalledTimes(1);
    });

    it('should log with some content', () => {
      logger.groupCollapsed = jest.fn();
      logger.groupEnd = jest.fn();
      logger.log = jest.fn();
      group.group('some title', { foo: 'bar' });
      expect(logger.groupCollapsed).toHaveBeenCalledTimes(1);
      expect(logger.groupEnd).toHaveBeenCalledTimes(1);
      expect(logger.log).toHaveBeenCalledTimes(1);
    });
  });
});
