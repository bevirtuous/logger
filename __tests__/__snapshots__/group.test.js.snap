// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`group getFormattedTime returns a formatted time string 1`] = `"00:38:38.396"`;

exports[`group maxKeysLength should return the highest length of a key in an object 1`] = `6`;

exports[`group pad returns a string of a number with one leading 0 1`] = `"5"`;

exports[`group pad returns a string of a number wuth three leading 0 1`] = `"0005"`;

exports[`group repeat returns the original string if no repeantance defined 1`] = `"foo"`;

exports[`group repeat returns the string 5 times 1`] = `"foofoofoofoofoo"`;

exports[`group style should return a valid styling with custom values. 1`] = `"color: #fff; font-weight: bold;"`;

exports[`group style should return a valid styling with default values. 1`] = `"color: gray; font-weight: lighter;"`;
